### DISCUSSION TITLE
Full DiscordNative support?

### DISCUSSION THREAD

**Discussion by KingGamingYT** [tsukasa/BdBrowser#44] [Ideas]
This may very well be out of the scope of this project, but I wanted to ask. Some plugins require DiscordNative modules for functionality.  These modules can be found in a local discord installation at `%LocalAppData%\Discord\app-(version)\modules`, however, I do not know if these same modules are accessible at all in the web version of Discord. Plugins that attempt to use native modules will fail to compile at startup. A particular example I'm pulling from has the error `Cannot read properties of undefined (reading 'requireModule')`. Below is the console output of `DiscordNative` on web compared to desktop:

Web
![image](https://github.com/user-attachments/assets/5d429449-13a2-4ae3-a13b-8b27e7def079)

Desktop
![image](https://github.com/user-attachments/assets/478af24a-475e-4531-a572-b9c7c604ae19)

While I don't fully understand the issue, I assume that the web browser itself takes over the tasks of most native modules and, as such, they aren't available.

**Comment by tsukasa** [comment]
Greetings KingGamingYT,

thank you for the question! Generally speaking it is unlikely that the contents of the desktop application's "modules" folder will be emulated or shimmed. They are absent from the web version due to native dependencies/functions either in Node/Electron or third-party (native) libraries. Today, BdBrowser provides a very small subset of some functions (i.e. for _very_ basic clipboard handling). More elaborate functions like OS processing handling, overlays, etc. are unlikely to happen due to the sandboxing of the web browser or because it does not make practical sense (in the case of the internal updater). 

If you have a specific plugin in mind, I'd be happy to take a look at it and check whether the used functions can be shimmed with reasonable effort in the browser - or whether I can add a workaround.

**Comment by KingGamingYT** [comment]
Sure! The specific plugin I was attempting to use when I encountered this issue is [VoiceMessages](https://github.com/riolubruh/VoiceMessages), which uses DiscordNative's `discord_voice` module to be able to record and process audio.

**Reply by tsukasa** [reply]
Greetings,

thank you for your patience, I finally had a chance to take a look at the VoiceMessages plugin. The tl;dr is: Getting the plugin to work on the web client requires some hammering, but seems possible enough.

Implementing the two required functions `startLocalAudioRecording` and `stopLocalAudioRecording` - even given Chrome does not support audio/ogg through its MediaRecorder API - is possible through Chrome's APIs and the existing Node shims. I have a buggy implementation in testing that I hope to stabilize enough for pushing. Right now, the implementation lacks error handling, records to a fixed filename and does not release the microphone when done.

Preliminary testing indicates that Discord also accepts audio/webm (even though the plugin tries to push the file as audio/ogg), so there is no need to polyfill the MediaRecorder. I cannot test this on mobile, but the desktop client plays the audio/webm file without issues.

I am somewhat concerned about how the plugin handles things internally, though.

The VoiceMessage Webpack search seems odd. The plugin performs the following search, which fails on the web client:
```javascript
Webpack.getModules(BdApi.Webpack.Filters.byKeys("Z")).filter(obj => obj.Z.type).filter(obj => obj.Z.type.toString().includes("waveform:"))[0].Z;
```
Instead, the exact same search without the "type" could be used instead, resulting in the desired results on both the desktop and web client:
```javascript
Webpack.getModules(BdApi.Webpack.Filters.byKeys("Z")).filter(obj => obj.Z).filter(obj => obj.Z.toString().includes("waveform:"));
```

Adapting the Webpack search allows the plugin to load on BdBrowser without startup errors. With the discord_voice module functions partially implemented, you can record and save an audio file to the VFS and pass the path back to the plugin through the callback.

The plugin performs some path cleanups that cause errors:
```javascript
filePath = filePath.replaceAll("/", "\\");
filePath = filePath.replaceAll(`\\`, `\\\\`);
```
This results in paths like `AppData\\BetterDiscord\\data\\recording.webm`, which currently breaks the VFS functions and results in a "file not found". I am still contemplating whether or not I want to support these "malformed" paths. In the meantime, one can remove the two instructions from the plugin as a workaround.

Sending the resulting file embeds the voice message as expected.

As soon as I am able to squash the bugs in the two discord_voice functions, I plan on pushing the changes so you can give it a test (i.e. whether mobile clients also accept the webm opus file).

**Reply by tsukasa** [reply]
Addendum: 1375eb6 contains support for discord_voice, [this artifact](https://github.com/tsukasa/BdBrowser/actions/runs/12338596541) can be used for testing.

**Reply by KingGamingYT** [reply]
The wait was worth it! I really appreciate the lengths you've gone to to support something so seldom used. audio/ogg is used because otherwise mobile versions of Discord, specifically iOS, won't play the audio due to Apple's restriction on allowed codecs on their devices :/. I've heard this restriction may have been circumvented in newer versions of the mobile app but since I use an older version I can't confirm that for sure.

Edit: After trying out the test build and making the changes to the plugin, here are some notes:
- Audio will **not** play on my version of mobile (186), it doesn't even attempt to do so
- Neither the preview of a recorded file or an uploaded file will appear in the record modal, it's just empty space
- Waveforms of recorded audio aren't traversable, and clicking anywhere on the waveform when playing will only pause the audio for however long you hold the mouse button and then continue playing (reminds me a lot of the webm-format video bugs Discord had a couple years ago, I wonder if this is related?)
- The download button for audio messages as an additional feature doesn't seem to work

**Reply by tsukasa** [reply]
Thank you for trying the mobile client and listing the issues with the current implementation! This feedback is incredibly helpful because I have no prior experience with that feature.

It stands to reason that the unwanted behaviours you describe are mostly due to uploading the correct codec in the wrong container format.

I will experiment to see if there is a neat way to achieve OGG+Opus support without having to go the WASM route.
